### ISSUE TITLE
Debouncing Field Validation

### ISSUE THREAD

**Issue by pmoeller91** [final-form/react-final-form#369]
### Are you submitting a **bug report** or a **feature request**?

Feature Request

### What is the current behavior?

Currently, there is no way I can find that would allow for debouncing field-level validation. Field-level validation is very nice, but rerunning the validation on every keystroke makes for a poor user experience and lower performance.

Async field-level validation also results in firing off numerous async requests with no trivial way to debounce that, either. Even the [linked example for asynchronous validation](https://codesandbox.io/s/wy7z7q5zx5) demonstrates this poor rapid-fire async behavior. (Select username field, select a different field, select username field again. Try typing: "Georgeee" at a moderate pace, and watch as the "username taken" message appears and rapidly disappears as the previous async validation returns and then is replaced by the next async validation)

Right now, even pausing the validation via the form API will result in react-final-form completely stopping all form changes until validation is resumed.

### What is the expected behavior?

The ability to debounce validation. All validation should optionally be delayed until a set interval has passed without the form changing. This will reduce the pressure caused my firing numerous async validations, reduce overhead associated with validating even one field on every render, and improve the user experience by delaying error display until typing is completed. Even at 60wpm validation will be triggered on a field as frequently as 5 times every single second.

### Other information

There was [another issue](https://github.com/final-form/react-final-form/issues/353) on this same topic, but the author closed the issue and received no responses.

I have searched pretty far and wide on this issue and have not found any satisfactory answer. Some people have attempted variations of it, but as far as I can tell none have succeeded. It seems like performing field-level validation at absolutely every single render/update is deeply built into final-form itself. There is absolutely no way right now for validation to separated cleanly from the process, and no way for errors to be delivered truly asynchronously using the built-in validation functions.

I could be wrong, and maybe I am! If so, please demonstrate how you can cause a field validation to fire only once when typing stops.

Using OnBlur validation could be one potential solution to help avoid this problem, but it seems more like a bandaid than a true fix.

**Comment by tkvw** [comment]
You can use `lodash/debounce` function, something like (untested code): 

```
import {Field} from "react-final-form";
import debounce from "lodash/debounce"
class DebouncingValidatingField extends React.Component{
   validate = debounce(this.props.validate,500)
   render(){
      return <Field {...this.props} validate={this.validate}/>
    }
}
```

**Comment by pmoeller91** [comment]
@tkvw A variation of that was among the very first things I tried. Passing a debounced function as the validation function causes validation to fail completely.
The validate function is expected to either: Immediately return a promise (which is then collected into an array, every single time validation is run), or immediately return either undefined or an error message. Either way, it requires that something be returned on every single validation pass, and every single validation pass must eventually resolve to either valid or invalid. There is no way to skip validation passes or delay them.

**Comment by tkvw** [comment]
@pmoeller91 : I see you're right: you can try this: 
```
import React from "react";
import PropTypes from "prop-types";
import { Field } from "react-final-form";

class DebouncingValidatingField extends React.Component {
  static propTypes = {
    debounce: PropTypes.number
  };
  static defaultProps = {
    debounce: 500
  };
  validate = (...args) =>
    new Promise(resolve => {
      if (this.clearTimeout) this.clearTimeout();
      const timerId = setTimeout(() => {
        resolve(this.props.validate(...args));
      }, this.props.debounce);
      this.clearTimeout = () => {
        clearTimeout(timerId);
        resolve();
      };
    });
  render() {
    return <Field {...this.props} validate={this.validate} />;
  }
}

export default DebouncingValidatingField;
```

@see: https://codesandbox.io/s/mmywp9jl1y

**Comment by l3v1k** [comment]
Thank you @tkvw, that is good solution!

But be aware from using Form "validating" flag on your debounced input, if you want to set eg. input disabled={true} while is's validating. You must provide own "validating" prop from DebouncingValidatingField. Don't use "validating" prop that Form provide, because it will trigger to **true** when your component "debouncing", so you will lose input focus when type one letter. 💪

**Comment by muyiwaoyeniyi** [comment]
@l3v1k Please can you explain a bit more what you're referring to? I'm have that problem but not sure how to resolve it. Thanks!

**Comment by xat** [comment]
If you don't want to implement the promise debounce logic yourself, you might want to checkout some utility library like this one: https://github.com/sindresorhus/p-debounce

**Comment by ziaulrehman40** [comment]
I am still struggling to achieve this debounce behavior on gloabl form level validation. As my form is huge and i don't want to implement per field validations( I am using yup to define schema and validate against it in `Form`'s validate).
I guess lodash's `debounce` does not play well with async/await or I am doing something wrong(probably)

**Comment by artegen** [comment]
We also have the case as @ziaulrehman40. 

Async validation examples seem to overlook this common case: when you have a form and need to validate async only one field. Not only in Final-forms, but also in Formik, React-hook-form etc. 

I think, Yup is a good declarative way to validate a form. So we prefer to use it, instead of duplicating validation rules for single fields. But when there's an async validation it breaks, because Yup can't pass loading state to form's library. On the other hand, separating one field for async validation from Yup, seems pushing you to not use Yup at all. 

So for now, fixing this common case on my own, the best (but not great) way I see, is to have validation function which tracks async field changes, validates it, then runs common validation with Yup, and then needs to merge these results and return in one object. 

But this still requires a bunch of state tracking happening around the React components.

I'd rather prefer to finally have a common solution to this use case.

**Comment by andtos90** [comment]
Thanks @artegen for your thoughts, I found the same issues and I'm now working on something related to your solution. Did you found any problem or can you share a snippet of your solution?

**Comment by luisgrases** [comment]
My solution was to use a mix of 'debounce-promise' + 'memoizee'

```
import memoize from 'memoizee'
import debounce from 'debounce-promise'

const debounced = debounce(async value => {
    const response = await client('GET', '/api-endpoint')

  const memoized = memoize(debounced, { promise: true, length: 1, max: 1 })
```

**Comment by bostrom** [comment]
Same scenario here as @ziaulrehman40 @artegen and @andtos90. Any suggestions?

**Comment by luisgrases** [comment]
@bostrom this is my solution:


```
import memoize from 'memoizee'

const createFormValidation = ({
  id,
  message,
  validationFn,
  async = false,
}: createFormValidationArgs) => {
  const memoized = memoize(validationFn, { length: 1 })
  const validationFnWrapper = async function(value) {
    if (!async) {
      return await validationFn.call(this, value)
    }
    return await memoized.call(this, value)
  }
  return [id, message, validationFnWrapper]
}
```

Create your validation:

```
 const validateUserName =  createFormValidation({
    id: 'checkUniqueUserName',
    message: "This user name is not valid",
    async: true,
    validationFn: async (value) =>  {
      const isValid = await validateNameAsynchronosly(value)
      return isValid
    },
  })
```

Usage:

```
name:  yup.string().test(...validateUserName)
```

**Comment by alx2das** [comment]
Interesting.
When we have many fields with errors and we write to one of them, at the moment of checking the error text is removed from all fields.
why is that?

[Example](https://codesandbox.io/s/react-final-form-asynchronous-field-level-debouncing-validation-example-forked-tc9j4?file=/index.js)

**Comment by onosendi** [comment]
Here's a memoized debounced version:

```js
function MemoizedDebouncedValidationField({
  milliseconds = 400,
  validate,
  ...props
}) {
  const timeout = useRef(null);
  const lastValue = useRef(null);
  const lastResult = useRef(null);

  const debounceValidation = (value, values, meta) => new Promise((resolve) => {
    if (timeout) {
      clearTimeout(timeout.current);
    }
    if (value !== lastValue.current) {
      timeout.current = setTimeout(() => {
        lastValue.current = value;
        lastResult.current = validate(value, values, meta);
        resolve(lastResult.current);
      }, milliseconds);
    } else {
      resolve(lastResult.current);
    }
  });

  return <Field validate={debounceValidation} {...props} />;
}
```

Usage:
```js
<MemoizedDebouncedValidationField
  name="username"
  validate={(value) => (value === 'jim' ? 'Username exists' : undefined)}
  render={({ input, meta }) => (
    <>
      <input {...input} />
      {(meta.touched && meta.error) && <p>Error</p>}
    </>
  )}
/>
```

**Comment by Radomir-Drukh** [comment]
It is 4 in the morning. I don't know why it works and don't care at this point. Here is a debounced validation function for the whole form:

`const timeout = useRef<(() => void) | null>(null);

    const validateForm = (values: any) =>
        new Promise((resolve) => {
            if (timeout.current) {
                timeout.current();
            }

            const timerId = setTimeout(() => {
                resolve(YOUR_VALIDATE_FUNC(values));
            }, DEBOUNCE_INTERVAL);

            timeout.current = () => {
                clearTimeout(timerId);
                resolve(true);
            };
        });
`
Later just pass it to form like 
`
<Form
  validate={validate form}
/>
`

**Comment by orenkole** [comment]
`
Possible solution for Field level async validation:

import { useRef, useCallback } from 'react';

function debounceAsync(fn, wait) {
  let timeout;
  let promiseResolver;
  return (...args) => {
    clearTimeout(timeout);
    return new Promise(resolve => {
      promiseResolver = resolve;
      timeout = setTimeout(async () => {
        const result = await fn(...args);
        promiseResolver(result);
      }, wait);
    });
  };
}

const debouncedValueCheck = debounceAsync(async (formValues, queryUrl) => {
  try {
    const queryResult = await axios.get(queryUrl);
    const checkResultMessage = getCheckResultMessage(queryResult);
    return checkResultMessage;
  } catch {
    console.error('Validation request failed');
  }
  return undefined;
}, 300);

export const useValueValidator = () => {
  const previousValue = useRef<string | undefined>();
  const previousError = useRef<string | undefined>();

  const validateValue = useCallback(
    async (value, allValues) => {
      if (value === previousValue.current) {
        return previousError.current;
      }

      previousValue.current = value;

      if (!value) {
        previousError.current = undefined;
        return undefined;
      }

      const error = await debouncedValueCheck({ ...allValues, value });
      lastError.current = error;
      return error;
    },
    []
  );

  return validateValue;
};

const validate = useValueValidator();

<Field
  validate={validate}
/>
`
